src/librustc_mir/interpret/eval_context.rs

```diff
@@ -557,56 +559,142 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+    pub(super) fn pop_stack_frame_internal(
+        &mut self,
+        unwinding: bool
+    ) -> InterpResult<'tcx, (StackPopCleanup, StackPopInfo)> {
+        info!("LEAVING({}) {} (unwinding = {})",
+            self.cur_frame(), self.frame().instance, unwinding);
+
         ::log_settings::settings().indentation -= 1;
         let frame = self.stack.pop().expect(
             "tried to pop a stack frame, but there were none",
         );
-        M::stack_pop(self, frame.extra)?;
+        let stack_pop_info = M::stack_pop(self, frame.extra)?;
+
         // Abort early if we do not want to clean up: We also avoid validation in that case,
         // because this is CTFE and the final value will be thoroughly validated anyway.
         match frame.return_to_block {
-            StackPopCleanup::Goto(_) => {},
-            StackPopCleanup::None { cleanup } => {
+            StackPopCleanup::Goto{ .. } => {},
+            StackPopCleanup::None { cleanup, .. } => {
+                assert!(!unwinding, "Encountered StackPopCleanup::None while unwinding");
+
                 if !cleanup {
                     assert!(self.stack.is_empty(), "only the topmost frame should ever be leaked");
                     // Leak the locals, skip validation.
-                    return Ok(());
+                    return Ok((frame.return_to_block, stack_pop_info));
                 }
             }
         }
         // Deallocate all locals that are backed by an allocation.
         for local in frame.locals {
             self.deallocate_local(local.value)?;
         }
-        // Validate the return value. Do this after deallocating so that we catch dangling
-        // references.
-        if let Some(return_place) = frame.return_place {
-            if M::enforce_validity(self) {
-                // Data got changed, better make sure it matches the type!
-                // It is still possible that the return place held invalid data while
-                // the function is running, but that's okay because nobody could have
-                // accessed that same data from the "outside" to observe any broken
-                // invariant -- that is, unless a function somehow has a ptr to
-                // its return place... but the way MIR is currently generated, the
-                // return place is always a local and then this cannot happen.
-                self.validate_operand(
-                    self.place_to_op(return_place)?,
-                    vec![],
-                    None,
-                )?;
+
+        // If we're popping frames due to unwinding, and we didn't just exit
+        // unwinding, we skip a bunch of validation and cleanup logic (including
+        // jumping to the regular return block specified in the StackPopCleanup)
+        let cur_unwinding = unwinding && stack_pop_info != StackPopInfo::StopUnwinding;
+
+        info!("StackPopCleanup: {:?} StackPopInfo: {:?} cur_unwinding = {:?}",
```

Comment:
This is probably to verbose to be always enabled for debugging. Sounds more like `trace!` to me.